### DISCUSSION TITLE
deisotopeAndSingleCharge

### DISCUSSION THREAD

**Discussion by anamaria-uofg** [OpenMS/OpenMS#7635] [Code]
Hello. I have just started using pyopenms. I wanted to ask whether there is a way to get the list of peaks for each isotopic cluster and not just the ion with the highest intensity after using deisotopeAndSingleCharge()? Or is there a another function which could do that? Thank you

**Comment by jpfeuffer** [comment]
Hi,
I have never used it myself but you could try with setting the keep_only_deisotoped parameter to False and investigate the FloatDataArrays of the modified spectrum afterwards.
Other than that, I cannot think of any function that does something similar right now.

You can deisotope it, generate Isotope patterns given the monoisotopes and then match the peaks to that. Not great.

**Reply by anamaria-uofg** [reply]
Hi, 
Thank you for your reply. I've just tried that, but I'm only getting an empty list when running getFloatDataArrays().

**Reply by jpfeuffer** [reply]
Did you activate the "annotateXYZ" parameters or how they are called?

**Reply by jpfeuffer** [reply]
https://github.com/OpenMS/OpenMS/blob/5a58f8fc83fcecb3a468679165bc4942a8cbcbe7/src/openms/include/OpenMS/PROCESSING/DEISOTOPING/Deisotoper.h#L104

**Reply by jpfeuffer** [reply]
Ah it's IntegerDataArray also

**Reply by jpfeuffer** [reply]
But reading the code, I am not sure if you can exactly get what you want (even if you have charges annotated. You would need to split charges and hope they are not consecutive in two neighboring clusters etc.). I agree it would be very helpful and I would feel limited, too, when using this function.
There should be an annotate cluster parameter.
Would be super helpful for visualizations.

Pinging @timosachsenberg who wrote the code initially

**Reply by anamaria-uofg** [reply]
Yes, that would be very helpful, thank you.
In the meantime I've tried looking at the IntegerDataArray. It seems that the second IntegerDataArray object, which I suppose comes from setting annotate_iso_peak_count = True, is only returning the number of isotopes per cluster.

<img width="736" alt="image" src="https://github.com/user-attachments/assets/8ac81427-9012-40e8-a775-9024247d9bb8">

**Reply by anamaria-uofg** [reply]
Hi,
Not sure if you've had a chance to have a look at adding this feature, but (with my super rusty c++ knowledge) I assume it would be done similarly to lines 269-272 in https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/PROCESSING/DEISOTOPING/Deisotoper.cpp ?
Initialise a vector mono_iso_peak_cluster and then just add the peak mz (instead of intensity) to each cluster mono_iso_peak_cluster[current_peak].push_back(old_spectrum[clusters[best_idx * max_isopeaks + i]].getMZ()); ?
Thanks again,
anamaria

**Reply by jpfeuffer** [reply]
Yes I think this will be the way to go. I am waiting on feedback from the codeowner @timosachsenberg

**Reply by timosachsenberg** [reply]
Sorry for the late response. The C++ code already contains logic to assign feature numbers to individual peaks. It is just not exported as IntegerDataArray.
What one would need to do:
- add a parameter to report feature numbers (bool annotate_feature)
- if annotate_feature is true: add a new IntegerDataArray that store the feature number for each peak (or -1 if not assigned)
something like: 
```
  spec.getIntegerDataArrays().back().setName("feature_number");
  feature_number_dataarray_index = spec.getIntegerDataArrays().size() - 1;
```
- and store the std::vector<Int> features(old_spectrum.size(), -1);  at the end of the function in this array:
```
  spec.getIntegerDataArrays()[feature_number_dataarray_index].swap(features);
```

**Reply by anamaria-uofg** [reply]
Thank you very much for your replies. I've created a pull request to add the code above, if that's okay with you.

**Comment by timosachsenberg** [comment]
PR has been integrated

**Comment by anamaria-uofg** [comment]
Hello. It's been a while, but I've only now tested this function in the updated version of pyopenms. It seems that now it is labelling just the deisotoped ions. I thought it would label all ions in order to indicate the clusters they belong to. Here is an example of what I am getting now. 

<img width="748" height="380" alt="image" src="https://github.com/user-attachments/assets/50376e10-7e6f-49f0-a3ea-77bad456ad0c" />

I was thinking that for feature number I would get a list with length equal to the m/z number before deisotoping looking something like this: feature_number = [1,1,1,1,2,2,2,2,3,1,2,-1,-1,2,4,4,....etc]. Would it be possible to achieve this easily with the current code? Thank you.
